File: android/app/src/main/java/com/kreativekoala/surfsense/LocalStorage.kt
package com.kreativekoala.surfsense

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

/**
 * Local storage manager for caching data and tracking usage history
 */
class LocalStorage private constructor(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val gson = Gson()

    companion object {
        private const val PREFS_NAME = "surfsense_local_storage"
        private const val KEY_CACHED_DEVICES = "cached_linked_devices"
        private const val KEY_CACHED_SUMMARIES = "cached_summaries"
        private const val KEY_LAST_FETCH_TIME = "last_fetch_time"
        private const val KEY_USAGE_HISTORY = "usage_history"
        private const val CACHE_TIMEOUT_MS = 5 * 60 * 1000L // 5 minutes

        @Volatile
        private var instance: LocalStorage? = null

        fun getInstance(context: Context): LocalStorage {
            return instance ?: synchronized(this) {
                instance ?: LocalStorage(context.applicationContext).also { instance = it }
            }
        }
    }

    // MARK: - Linked Devices Cache

    fun cacheLinkedDevices(devices: List<ClientInfo>) {
        val json = gson.toJson(devices)
        prefs.edit()
            .putString(KEY_CACHED_DEVICES, json)
            .putLong(KEY_LAST_FETCH_TIME, System.currentTimeMillis())
            .apply()
    }

    fun getCachedLinkedDevices(): List<ClientInfo>? {
        val json = prefs.getString(KEY_CACHED_DEVICES, null) ?: return null
        val type = object : TypeToken<List<ClientInfo>>() {}.type
        return try {
            gson.fromJson(json, type)
        } catch (e: Exception) {
            null
        }
    }

    fun clearCachedDevices() {
        prefs.edit().remove(KEY_CACHED_DEVICES).apply()
    }

    // MARK: - Usage Summary Cache

    fun cacheUsageSummary(summary: Map<String, Int>, deviceId: String, date: String) {
        val allSummaries = getAllCachedSummaries().toMutableMap()
        val key = "${deviceId}_$date"
        allSummaries[key] = summary
        prefs.edit().putString(KEY_CACHED_SUMMARIES, gson.toJson(allSummaries)).apply()
    }

    fun getCachedUsageSummary(deviceId: String, date: String): Map<String, Int>? {
        val allSummaries = getAllCachedSummaries()
        val key = "${deviceId}_$date"
        return allSummaries[key]
    }

    private fun getAllCachedSummaries(): Map<String, Map<String, Int>> {
        val json = prefs.getString(KEY_CACHED_SUMMARIES, null) ?: return emptyMap()
        val type = object : TypeToken<Map<String, Map<String, Int>>>() {}.type
        return try {
            gson.fromJson(json, type) ?: emptyMap()
        } catch (e: Exception) {
            emptyMap()
        }
    }

    // MARK: - Usage History (for trends)

    data class DailyUsage(
        val date: String,
        val totalMinutes: Int,
        val byCategory: Map<String, Int>
    )

    fun saveUsageHistory(usage: DailyUsage) {
        val history = getUsageHistory().toMutableList()

        // Remove old entry for same date if exists
        history.removeAll { it.date == usage.date }
        history.add(usage)

        // Keep only last 30 days, sorted by date descending
        val trimmedHistory = history
            .sortedByDescending { it.date }
            .take(30)

        prefs.edit().putString(KEY_USAGE_HISTORY, gson.toJson(trimmedHistory)).apply()
    }

    fun getUsageHistory(): List<DailyUsage> {
        val json = prefs.getString(KEY_USAGE_HISTORY, null) ?: return emptyList()
        val type = object : TypeToken<List<DailyUsage>>() {}.type
        return try {
            gson.fromJson(json, type) ?: emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun getUsageHistory(days: Int): List<DailyUsage> {
        return getUsageHistory().take(days)
    }

    // MARK: - Last Fetch Time

    fun getLastFetchTime(): Long {
        return prefs.getLong(KEY_LAST_FETCH_TIME, 0)
    }

    fun shouldRefresh(): Boolean {
        val lastFetch = getLastFetchTime()
        return System.currentTimeMillis() - lastFetch > CACHE_TIMEOUT_MS
    }

    // MARK: - Clear All Cache

    fun clearAllCache() {
        prefs.edit()
            .remove(KEY_CACHED_DEVICES)
            .remove(KEY_CACHED_SUMMARIES)
            .remove(KEY_LAST_FETCH_TIME)
            .remove(KEY_USAGE_HISTORY)
            .apply()
    }
}

File: ios/SurfSense/LocalStorage.swift
import Foundation
import os

private let logger = Logger(subsystem: "com.kreativekoala.SurfSense", category: "LocalStorage")

/// Local storage manager for caching data
class LocalStorage {
    static let shared = LocalStorage()

    private let defaults = UserDefaults.standard
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private init() {}

    // MARK: - Keys
    private enum Keys {
        static let cachedDevices = "cached_linked_devices"
        static let cachedUsageSummary = "cached_usage_summary"
        static let lastFetchTime = "last_fetch_time"
        static let usageHistory = "usage_history"
    }

    // MARK: - Linked Devices Cache

    func cacheLinkedDevices(_ devices: [ClientInfo]) {
        do {
            let data = try encoder.encode(devices)
            defaults.set(data, forKey: Keys.cachedDevices)
            defaults.set(Date(), forKey: Keys.lastFetchTime)
        } catch {
            logger.error("Failed to cache devices: \(error.localizedDescription)")
        }
    }

    func getCachedLinkedDevices() -> [ClientInfo]? {
        guard let data = defaults.data(forKey: Keys.cachedDevices) else { return nil }
        return try? decoder.decode([ClientInfo].self, from: data)
    }

    func clearCachedDevices() {
        defaults.removeObject(forKey: Keys.cachedDevices)
    }

    // MARK: - Usage Summary Cache

    func cacheUsageSummary(_ summary: [String: Int], for deviceId: String, date: String) {
        var allSummaries = getAllCachedSummaries()
        let key = "\(deviceId)_\(date)"
        allSummaries[key] = summary

        do {
            let data = try encoder.encode(allSummaries)
            defaults.set(data, forKey: Keys.cachedUsageSummary)
        } catch {
            logger.error("Failed to cache summary: \(error.localizedDescription)")
        }
    }

    func getCachedUsageSummary(for deviceId: String, date: String) -> [String: Int]? {
        let allSummaries = getAllCachedSummaries()
        let key = "\(deviceId)_\(date)"
        return allSummaries[key]
    }

    private func getAllCachedSummaries() -> [String: [String: Int]] {
        guard let data = defaults.data(forKey: Keys.cachedUsageSummary) else { return [:] }
        return (try? decoder.decode([String: [String: Int]].self, from: data)) ?? [:]
    }

    // MARK: - Usage History (for trends)

    struct DailyUsage: Codable {
        let date: String
        let totalMinutes: Int
        let byCategory: [String: Int]
    }

    func saveUsageHistory(_ usage: DailyUsage) {
        var history = getUsageHistory()

        // Remove old entry for same date if exists
        history.removeAll { $0.date == usage.date }
        history.append(usage)

        // Keep only last 30 days
        let sortedHistory = history.sorted { $0.date > $1.date }
        let trimmedHistory = Array(sortedHistory.prefix(30))

        do {
            let data = try encoder.encode(trimmedHistory)
            defaults.set(data, forKey: Keys.usageHistory)
        } catch {
            logger.error("Failed to save usage history: \(error.localizedDescription)")
        }
    }

    func getUsageHistory() -> [DailyUsage] {
        guard let data = defaults.data(forKey: Keys.usageHistory) else { return [] }
        return (try? decoder.decode([DailyUsage].self, from: data)) ?? []
    }

    func getUsageHistory(days: Int) -> [DailyUsage] {
        return Array(getUsageHistory().prefix(days))
    }

    // MARK: - Last Fetch Time

    func getLastFetchTime() -> Date? {
        return defaults.object(forKey: Keys.lastFetchTime) as? Date
    }

    func shouldRefresh(cacheTimeout: TimeInterval = 300) -> Bool {
        guard let lastFetch = getLastFetchTime() else { return true }
        return Date().timeIntervalSince(lastFetch) > cacheTimeout
    }

    // MARK: - Clear All Cache

    func clearAllCache() {
        defaults.removeObject(forKey: Keys.cachedDevices)
        defaults.removeObject(forKey: Keys.cachedUsageSummary)
        defaults.removeObject(forKey: Keys.lastFetchTime)
        defaults.removeObject(forKey: Keys.usageHistory)
    }
}
